File: animation/src/main/java/me/wyne/wutils/animation/ParallelAnimationStep.java
package me.wyne.wutils.animation;

import org.bukkit.Bukkit;
import org.javatuples.Pair;
import org.jetbrains.annotations.NotNull;

public class ParallelAnimationStep extends BaseAnimationStep {

    public ParallelAnimationStep(AnimationRunnable runnable, long delay, long period, long duration) {
        super(runnable, delay, period, duration);
    }

    public ParallelAnimationStep(AnimationRunnable runnable, long delay) {
        super(runnable, delay);
    }

    public ParallelAnimationStep(AnimationRunnable runnable) {
        super(runnable);
    }

    @Override
    protected void runOnce(@NotNull Animation animation) {
        var task = Bukkit.getScheduler().runTaskLater(
                animation.getPlugin(),
                () -> {
                    getRunnable().run(getDelay(), getPeriod(), getDuration());
                    close();
                    animation.getParallelTasks().remove(this);
        }, getDelay());
        animation.getParallelTasks().put(this, task);
        startNext(animation);
    }

    @Override
    protected void runRepeating(Animation animation) {
        var task = Bukkit.getScheduler().runTaskTimer(
                animation.getPlugin(),
                () -> {
                    if (getDuration() > 0 && ticksElapsed >= getDuration()) {
                        close();
                        var currentTask = animation.getParallelTasks().remove(this);
                        if (currentTask != null)
                            currentTask.cancel();
                        return;
                    }
                    getRunnable().run(getDelay(), getPeriod(), getDuration());
                    ticksElapsed += getPeriod();
        }, getDelay(), getPeriod());
        animation.getParallelTasks().put(this, task);
        startNext(animation);
    }

}

File: animation/src/main/kotlin/me/wyne/wutils/animation/ParallelAnimationStep.kt
package me.wyne.wutils.animation

import org.bukkit.Bukkit

class ParallelAnimationStep(
    runnable: AnimationRunnable,
    delay: Long = 0,
    period: Long = 0,
    duration: Long = 0
) : BaseAnimationStep(runnable, delay, period, duration) {
    override fun runOnce(animation: Animation) {
        Bukkit.getScheduler().runTaskLater(
            animation.plugin,
            { task ->
                animation.parallelTasks.put(this, task)
                Bukkit.getScheduler().runTask(animation.plugin, Runnable {
                    animation.pollStep()?.run(animation)
                })
                runnable.run(delay, period, duration)
                close()
                animation.parallelTasks.remove(this)
            },
            delay
        )
    }

    override fun runRepeating(animation: Animation) {
        Bukkit.getScheduler().runTaskTimer(
            animation.plugin,
            { task ->
                if (ticksElapsed <= 0) {
                    animation.parallelTasks.put(this, task)
                    Bukkit.getScheduler().runTask(animation.plugin, Runnable {
                        animation.pollStep()?.run(animation)
                    })
                }
                if (duration > 0 && ticksElapsed >= duration) {
                    close()
                    task.cancel()
                    animation.parallelTasks.remove(this)
                    return@runTaskTimer
                }
                runnable.run(delay, period, duration)
                ticksElapsed += period
            },
            delay,
            period
        )
    }
}
